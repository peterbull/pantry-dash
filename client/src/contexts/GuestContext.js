import { createContext, useState } from "react";

export const GuestContext = createContext();

export const GuestProvider = ({ children }) => {
  const [items, setItems] = useState([]);

  return (
    <GuestContext.Provider value={{ items, setItems }}>
      {children}
    </GuestContext.Provider>
  );
};
