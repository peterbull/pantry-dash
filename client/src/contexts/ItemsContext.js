/**
 * This module defines a context provider for items. It exports an ItemsContext object and an ItemsProvider component.
 * @module ItemsContext
 */

import React, { createContext, useState, useEffect } from "react";

/**
 * ItemsContext object created using createContext() method from React.
 * @type {object}
 * @property {array} items - An array of items.
 * @property {function} setItems - A function to set the items array.
 */
const ItemsContext = createContext();

/**
 * ItemsProvider component that provides the items context to its children.
 * @param {object} props - The component props.
 * @param {React.ReactNode} props.children - The child components to be wrapped by the ItemsProvider.
 * @returns {JSX.Element} - The JSX element representing the ItemsProvider component.
 */
const ItemsProvider = ({ children }) => {
  const [items, setItems] = useState([]);

  useEffect(() => {
    /**
     * Fetches the items from the server and sets the items state.
     * @async
     * @function fetchItems
     */
    const fetchItems = async () => {
      try {
        const response = await fetch(
          `http://${process.env.REACT_APP_DB_SERVER}:3001/items`
        );
        const data = await response.json();
        setItems(data);
      } catch (error) {
        console.error("Error fetching items:", error);
      }
    };

    fetchItems();
  }, []);

  return (
    <ItemsContext.Provider value={{ items, setItems }}>
      {children}
    </ItemsContext.Provider>
  );
};

export { ItemsContext, ItemsProvider };
