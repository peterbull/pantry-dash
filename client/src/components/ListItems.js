import React, { Fragment, useState, useEffect, useRef, useContext } from "react";
import EditItem from "./EditItem";
import CreateItem from "./CreateItem";
import { ItemsContext } from "../contexts/ItemsContext";
import SearchBar from "./SearchBar";

/**
 * Renders a table of pantry inventory items with the ability to add, edit, and delete items.
 * @returns {JSX.Element} The ListItems component.
 */
const ListItems = () => {
	// Get the items and setItems function from the context
	const { items, setItems } = useContext(ItemsContext);
	const [query, setQuery] = useState("");

	const listContainerRef = useRef(null);


	// Function to delete an item by ID
	const deleteItem = async (id) => {
		try {
			const deleteItem = await fetch(`http://localhost:5000/items/${id}`, {
				method: "DELETE"
			});

			if (deleteItem.ok) {
				// Update the items in the context
				setItems(prevItems => prevItems.filter(item => item.id !== id));
			}
		} catch (err) {
			console.log(err.message);
		}
	};

	// Function to handle the creation of a new item
	const handleItemCreated = (newItem) => {
		// Update the items in the context
		setItems(prevItems => [newItem, ...prevItems]);
	};

	// Filter items based on the search query
	const filteredItems = items.filter(item =>
		item.name.toLowerCase().includes(query.toLowerCase())
	);

	useEffect(() => {
		const currentScrollPosition = listContainerRef.current?.scrollTop;
		if (typeof currentScrollPosition === 'number') {
			listContainerRef.current.scrollTop = currentScrollPosition;
		}
	}, [filteredItems]);


	return (
		<Fragment>
			<div ref={listContainerRef} className="table-container">
				<h3 className="text-center">Pantry Inventory</h3>
				<SearchBar query={query} setQuery={setQuery} />
				<table className="table table-sm table-hover text-center responsive-sm">
					<thead>
						<tr>
							<th>Name</th>
							<th>Qty.</th>
							<th>Purchase More at</th>
						</tr>
					</thead>
					<tbody>
						<tr>
							{/* CreateItem component for adding new items */}
							<CreateItem onItemCreated={handleItemCreated} />
						</tr>
						{/* Loop through items and render EditItem component for each */}
						{filteredItems.map((item) =>
							<tr key={item.id}>
								<EditItem item={item} deleteItem={deleteItem} />
							</tr>
						)}
					</tbody>
				</table>
			</div>
		</Fragment>
	);
};

export default ListItems;
