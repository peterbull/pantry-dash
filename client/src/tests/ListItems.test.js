import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import { ItemsProvider } from "../contexts/ItemsContext";
import ListItems from '../components/ListItems';

describe('ListItems Component', () => {
  it('renders without crashing', () => {
    const { getByText } = render(
      <ItemsProvider>
        <ListItems />
      </ItemsProvider>
    );

    expect(getByText('Pantry Inventory')).toBeInTheDocument();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <ItemsProvider>
        <ListItems />
      </ItemsProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  })
});