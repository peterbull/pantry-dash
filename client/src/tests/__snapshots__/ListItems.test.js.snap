// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItems Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="table-container"
  >
    <h3
      class="text-center"
    >
      Pantry Inventory
    </h3>
    <input
      class="form-control mb-3"
      placeholder="Search..."
      type="text"
      value=""
    />
    <table
      class="table table-sm table-hover text-center responsive-sm"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Qty.
          </th>
          <th>
            Purchase More at
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <input
              class="form-control fw-light text-center"
              name="name"
              placeholder="Name"
              required=""
              type="text"
              value=""
            />
          </td>
          <td>
            <input
              class="form-control fw-light text-center"
              name="quantity"
              placeholder="Quantity"
              required=""
              type="number"
              value=""
            />
          </td>
          <td>
            <input
              class="form-control fw-light text-center"
              name="low_quantity"
              placeholder="Low Quantity"
              required=""
              type="number"
              value=""
            />
          </td>
          <td>
            <button
              class="btn btn-dark"
              type="submit"
            >
              Create
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
